### PR TITLE
VB-3330 Return all visits that are associated with the session templa…

### PR DESCRIPTION
**Pull request by AledWEvans**
To return closed count information

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt [hunk 1]
```diff
@@ -51,6 +51,21 @@ interface SessionTemplateRepository : JpaRepository<SessionTemplate, Long> {
     @Param("visitsToDate") visitsToDate: LocalDate?,
   ): Int
 
+  @Query(
+    "select count(*) from visit v " +
+      " JOIN session_template st ON st.reference = v.session_template_reference " +
+      " WHERE st.reference = :reference" +
+      " AND v.visit_start > :visitsFromDate" +
+      " AND (cast(:visitsToDate as date) is null OR v.visit_start < :visitsToDate)" +
+      " AND visit_status = 'CANCELLED' AND (v.outcome_status is null OR v.outcome_status != 'SUPERSEDED_CANCELLATION') ",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:60]: do migrated visits not have a outcome_status sometimes?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt [hunk 2]
```diff
@@ -51,6 +51,21 @@ interface SessionTemplateRepository : JpaRepository<SessionTemplate, Long> {
     @Param("visitsToDate") visitsToDate: LocalDate?,
   ): Int
 
+  @Query(
+    "select count(*) from visit v " +
+      " JOIN session_template st ON st.reference = v.session_template_reference " +
+      " WHERE st.reference = :reference" +
+      " AND v.visit_start > :visitsFromDate" +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:58]: should it be >= :visitsFromDate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt [hunk 3]
```diff
@@ -51,6 +51,21 @@ interface SessionTemplateRepository : JpaRepository<SessionTemplate, Long> {
...
+      " AND (cast(:visitsToDate as date) is null OR v.visit_start < :visitsToDate)" +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:59]: <= here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/SessionTemplateVisitStatsDto.kt [hunk 4]
```diff
@@ -10,10 +10,14 @@ data class SessionTemplateVisitStatsDto(
   @field:NotNull
   val minimumCapacity: SessionCapacityDto,
 
-  @Schema(description = "visit count for given date", example = "10", required = true)
+  @Schema(description = "visit counts", example = "10", required = true)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/SessionTemplateVisitStatsDto.kt:13]: is this the booked visit count?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:58]: This is tried and tested code... so we are already using this logic... will double check

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:59]: Yeha the visitToDate is exclusive so, buy looking at the code we have used this convention intermittently

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:60]: Possible,... tested failed with out the v.outcome_status is null

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:59]: looking at this we are gaining one more min thats all

**Review by dhirajshettymoj — APPROVED** [review]
Approved. As discussed the +1 day will be picked as part of a separate technical task.

**Pull request merged**